clients/redis/command.go

```diff
@@ -0,0 +1,308 @@
+)
+
+type CommandClient struct {
+	messageBus      messaging.MessageClient
+	queryMessages   chan types.MessageEnvelope
+	queryErrors     chan error
+	commandMessages chan types.MessageEnvelope
+	commandErrors   chan error
+	topics          map[string]string
+	timeout         time.Duration
+}
+
+func NewCommandClient(messageBus messaging.MessageClient, topics map[string]string, timeout time.Duration) (interfaces.CommandClient, error) {
+	client := &CommandClient{
+		messageBus: messageBus,
+		topics:     topics,
+		timeout:    timeout,
+	}
+
+	queryResponseTopic, ok := topics[QueryResponseTopic]
+	if ok {
+		queryMessages := make(chan types.MessageEnvelope, 1)
+		queryErrors := make(chan error)
+		queryTopics := []types.TopicChannel{
+			{
+				Topic:    queryResponseTopic,
+				Messages: queryMessages,
+			},
+		}
+		err := messageBus.Subscribe(queryTopics, queryErrors)
+		if err != nil {
+			return nil, err
+		}
+
+		client.queryMessages = queryMessages
+		client.queryErrors = queryErrors
+	}
+
+	commandResponseTopic, ok := topics[CommandResponseTopic]
+	if ok {
+		commandMessages := make(chan types.MessageEnvelope, 1)
+		commandErrors := make(chan error)
+		commandTopics := []types.TopicChannel{
+			{
+				Topic:    commandResponseTopic,
+				Messages: commandMessages,
+			},
+		}
+		err := messageBus.Subscribe(commandTopics, commandErrors)
+		if err != nil {
+			return nil, err
+		}
+
+		client.commandMessages = commandMessages
+		client.commandErrors = commandErrors
+	}
+
+	return client, nil
+}
+
+func (c *CommandClient) AllDeviceCoreCommands(ctx context.Context, offset int, limit int) (responses.MultiDeviceCoreCommandsResponse, error) {
+	if c.queryMessages == nil {
+		return responses.MultiDeviceCoreCommandsResponse{}, errors.New("commandquery request/response topics not provided")
+	}
+
+	queryParams := map[string]string{common.Offset: strconv.Itoa(offset), common.Limit: strconv.Itoa(limit)}
+	requestEnvelope := types.NewMessageEnvelopeForRequest(nil, queryParams)
+	requestTopic := strings.Join([]string{c.topics[QueryRequestTopic], common.All}, "/")
+	err := c.messageBus.Publish(requestEnvelope, requestTopic)
+	if err != nil {
+		return responses.MultiDeviceCoreCommandsResponse{}, err
+	}
+
+	select {
+	case <-ctx.Done():
+		return responses.MultiDeviceCoreCommandsResponse{}, nil
+	case <-time.After(c.timeout):
+		return responses.MultiDeviceCoreCommandsResponse{}, errors.New("timed out waiting response")
+	case err = <-c.queryErrors:
+		return responses.MultiDeviceCoreCommandsResponse{}, err
+	case responseEnvelope := <-c.queryMessages:
+		if responseEnvelope.ErrorCode == 1 {
+			return responses.MultiDeviceCoreCommandsResponse{}, errors.New(string(responseEnvelope.Payload))
+		}
+
+		var res responses.MultiDeviceCoreCommandsResponse
+		err = json.Unmarshal(responseEnvelope.Payload, &res)
+		if err != nil {
+			return responses.MultiDeviceCoreCommandsResponse{}, err
+		}
+
+		return res, nil
```

Comment:
We need to handle the case that we received a response that isn't for our request. i.e. need to match the request ID. If doesn't match, loop back up for the next message. Same for all APIs here.

This could happen if two separate App services are making command requests at the same time. 

This is good first implementation, but we might want to rethink we handle publishing and subscribing the responses by having the request Id in the response topic and subscribe to that very specific topic each time, rather then the general response topic that will cause all service to receive at response message and have to filter out the ones that are not theirs. Would need to unsubscribe after receiving or timeout?? Maybe some refactoring to consider for 3.0, since we are so close to code freeze.